### ISSUE TITLE
loss is very huge at the beginning and then becomes Non

### ISSUE THREAD

**Issue by tripmasterwang** [Chaos96/PointNSP#3]
Thank you for your creative work! When I reproduce your code, I meet following issue:

the loss is very huge at the beginning and then becomes Non. How can I figure it out?

Total number of data:2458
Min number of points: (train)2048 (test)2048
Total number of data:352
Min number of points: (train)2048 (test)2048
Loss: 3.4705, 
Loss: 41731.0056, 
Loss: 980185.6306, 
Loss: 1539726.5681, 
Loss: 1998357.3181, 
Loss: 11737154.3181, 
Loss: 32406446.3181, 
Loss: 44479200.3181, 
Loss: 53520095.3181, 
Loss: 74370883.3181, 
Loss: 491376451.3181, 
Loss: 22828625731.3181, 
Loss: 89964527300419.3125, 
Loss: 562254968860483.3125, 
Loss: 14005225579949892.0000, 
Loss: 64983644115119936.0000, 
Loss: 107792188744740672.0000, 
Loss: 109181175799557952.0000, 
Loss: 109233702930820928.0000, 
Loss: 109239841558966080.0000, 
Loss: 109240787645706048.0000, 
Loss: 109242861140782912.0000, 
Loss: 110537578763726656.0000, 
Loss: 124554998029430592.0000, 
Loss: 124640596081717056.0000, 
Loss: 153562345359748928.0000, 
Loss: 153772327468478272.0000, 
Loss: 166686132338896704.0000, 
Loss: 166709702035189568.0000, 
Loss: 168008659461943104.0000, 
Loss: 168035013660192576.0000, 
Loss: 168426011712237376.0000, 
Loss: 168445322468208448.0000, 
Loss: 168475852760887104.0000, 
Loss: 168540025247182656.0000, 
Loss: 168690313232854848.0000, 
Loss: 168820539183017792.0000, 
Loss: 168896070595325760.0000, 
Loss: 168949386905857856.0000, 
Loss: 168989431717113664.0000, 
Loss: 169076740634390336.0000, 
Loss: 169841433523203904.0000, 
Loss: 169984443250585408.0000, 
Loss: 170047458171891520.0000, 
Loss: 170161617152716608.0000, 
Loss: 170292121319452480.0000, 
Loss: 170434649658175296.0000, 
Loss: 170604417333800768.0000, 
Loss: 170763260575559488.0000, 
Loss: 170907040796585792.0000, 
Loss: 171136784502828864.0000, 
Loss: 171321803305332544.0000, 
Loss: 171441574927609664.0000, 
Loss: 171705198275407680.0000, 
Loss: 172139572410134336.0000, 
Loss: 172454221009596224.0000, 
Loss: 172757792821292864.0000, 
Loss: 6383767586488397824.0000, 
Loss: nan, 
Loss: nan, 
Loss: nan, 
Loss: nan, 
Loss: nan, 
Loss: nan, 
Loss: nan, 
Loss: nan, 
Loss: nan,

**Comment by zq-meng** [comment]
> Thank you for your creative work! When I reproduce your code, I meet following issue:
> 
> the loss is very huge at the beginning and then becomes Non. How can I figure it out?
> 
> Total number of data:2458 Min number of points: (train)2048 (test)2048 Total number of data:352 Min number of points: (train)2048 (test)2048 Loss: 3.4705, Loss: 41731.0056, Loss: 980185.6306, Loss: 1539726.5681, Loss: 1998357.3181, Loss: 11737154.3181, Loss: 32406446.3181, Loss: 44479200.3181, Loss: 53520095.3181, Loss: 74370883.3181, Loss: 491376451.3181, Loss: 22828625731.3181, Loss: 89964527300419.3125, Loss: 562254968860483.3125, Loss: 14005225579949892.0000, Loss: 64983644115119936.0000, Loss: 107792188744740672.0000, Loss: 109181175799557952.0000, Loss: 109233702930820928.0000, Loss: 109239841558966080.0000, Loss: 109240787645706048.0000, Loss: 109242861140782912.0000, Loss: 110537578763726656.0000, Loss: 124554998029430592.0000, Loss: 124640596081717056.0000, Loss: 153562345359748928.0000, Loss: 153772327468478272.0000, Loss: 166686132338896704.0000, Loss: 166709702035189568.0000, Loss: 168008659461943104.0000, Loss: 168035013660192576.0000, Loss: 168426011712237376.0000, Loss: 168445322468208448.0000, Loss: 168475852760887104.0000, Loss: 168540025247182656.0000, Loss: 168690313232854848.0000, Loss: 168820539183017792.0000, Loss: 168896070595325760.0000, Loss: 168949386905857856.0000, Loss: 168989431717113664.0000, Loss: 169076740634390336.0000, Loss: 169841433523203904.0000, Loss: 169984443250585408.0000, Loss: 170047458171891520.0000, Loss: 170161617152716608.0000, Loss: 170292121319452480.0000, Loss: 170434649658175296.0000, Loss: 170604417333800768.0000, Loss: 170763260575559488.0000, Loss: 170907040796585792.0000, Loss: 171136784502828864.0000, Loss: 171321803305332544.0000, Loss: 171441574927609664.0000, Loss: 171705198275407680.0000, Loss: 172139572410134336.0000, Loss: 172454221009596224.0000, Loss: 172757792821292864.0000, Loss: 6383767586488397824.0000, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan,

Dear,

Apologies for the delayed response, as we have been busy conducting large-scale experiments on the Objaverse dataset. May we ask which training stage your reported loss corresponds to—the first-stage multi-scale VQ-VAE training or the second-stage autoregressive (AR) training?

Best regards.

**Comment by zq-meng** [comment]
> Thank you for your creative work! When I reproduce your code, I meet following issue:
> 
> the loss is very huge at the beginning and then becomes Non. How can I figure it out?
> 
> Total number of data:2458 Min number of points: (train)2048 (test)2048 Total number of data:352 Min number of points: (train)2048 (test)2048 Loss: 3.4705, Loss: 41731.0056, Loss: 980185.6306, Loss: 1539726.5681, Loss: 1998357.3181, Loss: 11737154.3181, Loss: 32406446.3181, Loss: 44479200.3181, Loss: 53520095.3181, Loss: 74370883.3181, Loss: 491376451.3181, Loss: 22828625731.3181, Loss: 89964527300419.3125, Loss: 562254968860483.3125, Loss: 14005225579949892.0000, Loss: 64983644115119936.0000, Loss: 107792188744740672.0000, Loss: 109181175799557952.0000, Loss: 109233702930820928.0000, Loss: 109239841558966080.0000, Loss: 109240787645706048.0000, Loss: 109242861140782912.0000, Loss: 110537578763726656.0000, Loss: 124554998029430592.0000, Loss: 124640596081717056.0000, Loss: 153562345359748928.0000, Loss: 153772327468478272.0000, Loss: 166686132338896704.0000, Loss: 166709702035189568.0000, Loss: 168008659461943104.0000, Loss: 168035013660192576.0000, Loss: 168426011712237376.0000, Loss: 168445322468208448.0000, Loss: 168475852760887104.0000, Loss: 168540025247182656.0000, Loss: 168690313232854848.0000, Loss: 168820539183017792.0000, Loss: 168896070595325760.0000, Loss: 168949386905857856.0000, Loss: 168989431717113664.0000, Loss: 169076740634390336.0000, Loss: 169841433523203904.0000, Loss: 169984443250585408.0000, Loss: 170047458171891520.0000, Loss: 170161617152716608.0000, Loss: 170292121319452480.0000, Loss: 170434649658175296.0000, Loss: 170604417333800768.0000, Loss: 170763260575559488.0000, Loss: 170907040796585792.0000, Loss: 171136784502828864.0000, Loss: 171321803305332544.0000, Loss: 171441574927609664.0000, Loss: 171705198275407680.0000, Loss: 172139572410134336.0000, Loss: 172454221009596224.0000, Loss: 172757792821292864.0000, Loss: 6383767586488397824.0000, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan,

If the loss explosion occurs during the second-stage autoregressive (AR) training, it may be due to insufficient training in the first stage. Based on our experience, accurate and stable reconstruction learning in the first stage is critical for effective second-stage training.

If the loss explosion occurs in the first stage, we have also encountered this issue during our research. The stability of first-stage training is sensitive to many hyperparameters, such as numerical precision (dtype), batch size, and others. Even noted in the LION GitHub repository, the first-stage VAE can easily encounter NaN issues. Even to date, VAR has not released a well-trained checkpoint for first-stage codebook learning.

To help alleviate these issues, we will be updating the codebase shortly, as we have just completed large-scale experiments. The updated release will include corresponding checkpoints and detailed implementation information to facilitate reliable reproduction of our results. Thank you for your attention and kind understanding.

**Comment by zq-meng** [comment]
> Thank you for your creative work! When I reproduce your code, I meet following issue:
> 
> the loss is very huge at the beginning and then becomes Non. How can I figure it out?
> 
> Total number of data:2458 Min number of points: (train)2048 (test)2048 Total number of data:352 Min number of points: (train)2048 (test)2048 Loss: 3.4705, Loss: 41731.0056, Loss: 980185.6306, Loss: 1539726.5681, Loss: 1998357.3181, Loss: 11737154.3181, Loss: 32406446.3181, Loss: 44479200.3181, Loss: 53520095.3181, Loss: 74370883.3181, Loss: 491376451.3181, Loss: 22828625731.3181, Loss: 89964527300419.3125, Loss: 562254968860483.3125, Loss: 14005225579949892.0000, Loss: 64983644115119936.0000, Loss: 107792188744740672.0000, Loss: 109181175799557952.0000, Loss: 109233702930820928.0000, Loss: 109239841558966080.0000, Loss: 109240787645706048.0000, Loss: 109242861140782912.0000, Loss: 110537578763726656.0000, Loss: 124554998029430592.0000, Loss: 124640596081717056.0000, Loss: 153562345359748928.0000, Loss: 153772327468478272.0000, Loss: 166686132338896704.0000, Loss: 166709702035189568.0000, Loss: 168008659461943104.0000, Loss: 168035013660192576.0000, Loss: 168426011712237376.0000, Loss: 168445322468208448.0000, Loss: 168475852760887104.0000, Loss: 168540025247182656.0000, Loss: 168690313232854848.0000, Loss: 168820539183017792.0000, Loss: 168896070595325760.0000, Loss: 168949386905857856.0000, Loss: 168989431717113664.0000, Loss: 169076740634390336.0000, Loss: 169841433523203904.0000, Loss: 169984443250585408.0000, Loss: 170047458171891520.0000, Loss: 170161617152716608.0000, Loss: 170292121319452480.0000, Loss: 170434649658175296.0000, Loss: 170604417333800768.0000, Loss: 170763260575559488.0000, Loss: 170907040796585792.0000, Loss: 171136784502828864.0000, Loss: 171321803305332544.0000, Loss: 171441574927609664.0000, Loss: 171705198275407680.0000, Loss: 172139572410134336.0000, Loss: 172454221009596224.0000, Loss: 172757792821292864.0000, Loss: 6383767586488397824.0000, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan, Loss: nan,

Dear,

We have released all codes and uploaded checkpoints to Hugging Face.

Best Regards.
